test/test_misc_utils.py

```diff
@@ -293,6 +294,107 @@ def test_virtual_app_patch_json():
         ]
 
 
+def test_virtual_app_error():
+
+    error_message = "You did a bad thing."
+    offending_url = "http://fixture.4dnucleome.org/offending/url"
+    body_text = '{"alpha": "omega"}'
+    body_json = {"alpha": "omega"}
+    wrapped_error = Exception("Some other exception")
+
+    e = VirtualAppError(error_message, offending_url, body_text, wrapped_error)
+    m = str(e)
+
+    assert error_message in m
+    assert offending_url in m
+    assert body_text in m
+    assert str(wrapped_error) in m
+
+    # And the repr is the same.
+    assert repr(e) == str(e)
+
+    # NOTE: Weirdly, I think we'd have had complete code coverage even without this next test, but that illustrates
+    #  why code coverage counts aren't always the right metric. With different data, the same code paths sometimes
+    #  does different things in ways that code coverage tools don't register. -kmp 21-May-2020
+
+    e2 = VirtualAppError(error_message, offending_url, body_json, wrapped_error)
+    m2 = str(e2)
+
+    assert error_message in m2
+    assert offending_url in m2
+    assert str(body_json) in m2             # body_json will be rendered as a Python dict (e.g., with single quotes)
+    assert not json.dumps(body_json) in m2  # So body_json will NOT be rendered as JSON via json.dumps
+    assert str(wrapped_error) in m2
+
+    # And the repr is the same.
+    assert repr(e2) == str(e2)
+
+
+def test_virtual_app_crud_failure():
+
+    simulated_error_message = "simulated error"
+
+    class FakeTestApp(VirtualAppError):
```

Comment:
Why does this inherit from `VirtualAppError`?

Reply:
As noted over Zoom, this was various errors conspiring -- really the parent class didn't matter, so I've removed that. The fact that there was an unrelated point made about this error class in a comment farther down was just coincidence. I think since we talked we're all set on this.

Reply:
Same buggy issue now fixed. In the case of the second test, I just removed it as redundant. The crud test tests all three of `get`, `post_json`, and `patch_json`.